sql: implement REASSIGN OWNED BY ... TO ...

angelapwen opened the pull request:
Resolves #52022. 

The command should change ownerships of all database objects in the
current database owned by any roles in the first argument to the new role
in second argument. It requires the current user be a member of
both source role(s) and target role.

Release note (sql change): Implement the REASSIGN OWNED BY ... TO
... command, which changes ownership of all database objects in
the current database, owned by any roles in the first argument, to the
new role in the second argument.

Commit:
sql: implement REASSIGN OWNED BY ... TO ...
The command should change ownerships of all database objects in the
current database owned by any roles in the first argument to the new role
in second argument. It requires the current user be a member of
both source role(s) and target role.

Release note (sql change): Implement the REASSIGN OWNED BY ... TO
... command, which changes ownership of all database objects in
the current database, owned by any roles in the first argument, to the
new role in the second argument.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54594)
<!-- Reviewable:end -->

angelapwen commented on pkg/sql/reassign_owned_by.go:
```diff
@@ -13,15 +13,114 @@ package sql
+
+	// Check all roles in old roles exist
+	// And that current user has membership on old role (PostgresQL requirement)
+	// TODO(angelaw): Refactor the error checking out to authorization?
```
Does this seem like checking that should be done in the `authorization.go` file?

angelapwen commented on pkg/sql/reassign_owned_by.go:
```diff
@@ -13,15 +13,114 @@ package sql
+
+	// TODO(angelaw): This is messy but we need an immutableDbDesc for description and
+	// we only have a ResolveMutableDatabaseDescriptor — any better ideas?
+	lCtx := newInternalLookupCtx(ctx, allDescs, currentDbDesc.ImmutableCopy().(*dbdesc.Immutable))
```
See comment in code

angelapwen commented on pkg/sql/reassign_owned_by.go:
```diff
@@ -13,15 +13,114 @@ package sql
+	}
+
+	// Not sure we can directly call those methods as checkCanAlterToNewOwner is called and has different checks
+	// I think it actually works fine as long as we pass in hasOwnership = true
```
Would appreciate a gut check here on whether it's okay that we call methods that call https://github.com/cockroachdb/cockroach/blob/0b105ec147985b65a2331784d8aadc0a1eec44cb/pkg/sql/authorization.go#L552 (`checkCanAlterToNewOwner`) with a `true` value for `hasOwnership` and not affect functionality

angelapwen commented on pkg/sql/reassign_owned_by.go:
```diff
@@ -13,15 +13,114 @@ package sql
+		for _, dbDesc := range lCtx.dbDescs {
+			// We only expect 1 matching db (current database)
+			if p.IsOwner(dbDesc, oldRole) {
+				p.AlterDatabaseOwner(ctx, tree.AlterDatabaseOwner) // Does not include new owner string
```
Calling all of these methods, `AlterDatabaseOwner` `alterSchemaOwner, etc are not as straightforward as they are in psql. Each takes in different arguments that I will need to populate individually. 

Does this seem like the right path to keep moving towards, or should I look to directly change the ownership here without calling these methods?

arulajmani reviewed (changes requested):
Left a partial review after the first pass, so that you're unblocked. I'll come back to this and finish it. In particular, you need to split your code into a planning and execution phase. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MHlhc9o1swTqXJ9UQSk:b-uw21b0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @arulajmani, and @solongordon)

___
*[pkg/sql/authorization.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlNOzX7TApe42ke_fT:-MHlNOzX7TApe42ke_fU:b-8aoyi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/authorization.go#L159)):*
> ```Go
> 
> // IsOwner returns if the role has ownership on the descriptor.
> func (p *planner) IsOwner(desc catalog.Descriptor, role string) bool {
> ```

Is this required? 
___
*[pkg/sql/reassign_owned_by.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlX0KH31xm6ubJT_V_:-MHlX0KH31xm6ubJT_Va:bso2jvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L29)):*
> ```Go
> }
> 
> func (p *planner) ReassignOwnedBy(ctx context.Context, n *tree.ReassignOwnedBy) (planNode, error) {
> ```

A bulk of what you're doing here should happen in the execution phase. 
___
*[pkg/sql/reassign_owned_by.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-33:-MHlPdvk3p21iGJtqz_2:bz2rnsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L33)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Does this seem like checking that should be done in the `authorization.go` file?
</blockquote></details>

I think it's fine to have this checking here, as it's a series of steps that are specific to `REASSIGN OWNED`
___
*[pkg/sql/reassign_owned_by.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlSk6J1y6nztK0EgA7:-MHlSk6J1y6nztK0EgA8:bb7ks7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L82)):*
> ```Go
> 
> 	currentDatabase := p.CurrentDatabase()
> 	currentDbDesc, err := p.ResolveMutableDatabaseDescriptor(ctx, currentDatabase, true)
> ```

nit: When passing a value like true, we generally comment the parameter name. In this case, `/* required */`
___
*[pkg/sql/reassign_owned_by.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-97:-MHlUyhTAX8wA7SgVL7F:b-qdfwa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L97)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Calling all of these methods, `AlterDatabaseOwner` `alterSchemaOwner`, etc are not as straightforward as they are in psql. Each takes in different arguments that I will need to populate individually. 

Does this seem like the right path to keep moving towards, or should I look to directly change the ownership here without calling these methods? 
</blockquote></details>

I don't think this is doing what you want to do actually. The method you are calling (and plan to call) only create the "plan node". At a high level, the way this works is we have a planning phase, which constructs plan nodes, and an execution phase, where `startExec` is called on these plan nodes. By calling this method, you're simply creating the plan node, but because you never execute it, nothing is really happening! In fact, your code should follow similar structure and a lot of what you're doing here should be moved to the execution phase. You can look at any other file for inspiration on how to structure this -- I had `drop_role` open, so I'll point you to that :) 

The general strategy here would be to lift the stuff in `startExec` of `AlterDatabaseOwner` etc. into separate functions so they can be called from different places. You'll also need to make sure that the checking that happens in the planning phase is also valid/is bypassed on purpose when you call them from here. 
___
*[pkg/sql/reassign_owned_by.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlU1BV4E-_97WgsDRd:-MHlU1BV4E-_97WgsDRe:b10a9sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L102)):*
> ```Go
> 		for _, schemaDesc := range lCtx.schemaDescs {
> 			if p.IsOwner(schemaDesc, oldRole) {
> 				// What is job desc here?
> ```

I think it means  job description. I find the name quite confusing, while we're here we should rename the parameter to "jobDescription" instead in `alterSchemaOwner`. 


<!-- Sent from Reviewable.io -->

solongordon reviewed (commented):
Oops, I wrote some comments earlier and never clicked submit. Hopefully Arul and I agree with each other.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MHlsoUbBeQMcxcSY52i:bt59edr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen and @solongordon)

___
*[pkg/sql/reassign_owned_by.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-33:-MHlMR1j0tk5pTG9y0xK:b1btx97) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L33)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Does this seem like checking that should be done in the `authorization.go` file?
</blockquote></details>

I think it's reasonable to keep it here unless you're duplicating logic that's used elsewhere.
___
*[pkg/sql/reassign_owned_by.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHl_zl12mBRtmL7TF33:-MHl_zl12mBRtmL7TF34:b-tuanua) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L73)):*
> ```Go
> 			// TODO(angelaw): find error message from psql
> 			pgcode.InsufficientPrivilege, "current user must be member of role %q", n.NewRole)
> 	}
> ```

I'm curious if all these upfront checks you're doing are necessary or if they are already covered by privilege checks in the individual ALTER OWNER commands. Don't worry about ditching them yet but something to keep in mind.
___
*[pkg/sql/reassign_owned_by.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-89:-MHlYnsVCejFOv0zkOCK:b-ooot2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L89)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

See comment in code
</blockquote></details>

I actually wonder if we want to restrict this to the current database or not. Unlike Postgres we do not have a hard separation between databases. It might be better to just pass `nil` here and iterate over all databases.
___
*[pkg/sql/reassign_owned_by.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-97:-MHlWAV8Bd_lQhdmS3FW:bkdkrua) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L97)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Calling all of these methods, `AlterDatabaseOwner` `alterSchemaOwner`, etc are not as straightforward as they are in psql. Each takes in different arguments that I will need to populate individually. 

Does this seem like the right path to keep moving towards, or should I look to directly change the ownership here without calling these methods? 
</blockquote></details>

I *think* this is the right path since it looks like the implementations are not all analogous. (For instance the schema implementation creates a special schema change job which the others don't.) Once you have it working maybe we can revisit whether there are redundant checks happening between calls.

However it's kind of janky if you need to construct and pass in a `tree` node. I'd refactor such methods to take explicit parameters. For instance I would change the `alterTableOwner` params from `(ctx context.Context, n *alterTableNode, newOwner string)` to `(ctx context.Context, desc *tabledesc.Mutable, hasOwnership bool, newOwner string)`.
___
*[pkg/sql/reassign_owned_by.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-123:-MHlZkTJ0A3d-XvVEmsa:b-c30ews) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L123)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Would appreciate a gut check here on whether it's okay that we call methods that call https://github.com/cockroachdb/cockroach/blob/0b105ec147985b65a2331784d8aadc0a1eec44cb/pkg/sql/authorization.go#L552 (`checkCanAlterToNewOwner`) with a `true` value for `hasOwnership` and not affect functionality
</blockquote></details>

I'm not quite following here, can you say more?


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @arulajmani, and @solongordon)

___
*[pkg/sql/authorization.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlNOzX7TApe42ke_fT:-MHmR-p39bPNYtU1MULB:bxq7hce) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/authorization.go#L159)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this required? 
</blockquote></details>

I chatted with Solon and it sounds like I will be able to rely on the calls to checkCanAlterToNewOwner() nested within each of the Alter * Owner commands, so I may not need this, but just curious:

I had thought I needed to add the planner receiver in order to call IsOwner() directly from sql/reassign_owned_by. Is there a way to access this function elsewhere without adding the receiver? 
___
*[pkg/sql/reassign_owned_by.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlSk6J1y6nztK0EgA7:-MHmR_xS6epT6jtxbRTg:b226wtj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L82)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: When passing a value like true, we generally comment the parameter name. In this case, `/* required */`
</blockquote></details>

Thanks for letting me know! Will do
___
*[pkg/sql/reassign_owned_by.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-89:-MHme8o1864zjCNdpNy9:b-tepmc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L89)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I actually wonder if we want to restrict this to the current database or not. Unlike Postgres we do not have a hard separation between databases. It might be better to just pass `nil` here and iterate over all databases.
</blockquote></details>

Great, we'll go with this.
___
*[pkg/sql/reassign_owned_by.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHYU341iS7v4DU-L9u7-r1-97:-MHmeELZC2X8ReKAvl0M:b-axz3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L97)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I *think* this is the right path since it looks like the implementations are not all analogous. (For instance the schema implementation creates a special schema change job which the others don't.) Once you have it working maybe we can revisit whether there are redundant checks happening between calls.

However it's kind of janky if you need to construct and pass in a `tree` node. I'd refactor such methods to take explicit parameters. For instance I would change the `alterTableOwner` params from `(ctx context.Context, n *alterTableNode, newOwner string)` to `(ctx context.Context, desc *tabledesc.Mutable, hasOwnership bool, newOwner string)`.
</blockquote></details>

Thank you both! Chatted with Solon and understood the difference between the planning and execution phase better. 
___
*[pkg/sql/reassign_owned_by.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlU1BV4E-_97WgsDRd:-MHmRlBh3NgXtFfDOui_:bhy9ost) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L102)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think it means  job description. I find the name quite confusing, while we're here we should rename the parameter to "jobDescription" instead in `alterSchemaOwner`. 
</blockquote></details>

Will do!


<!-- Sent from Reviewable.io -->

angelapwen commented:
**Updates:** most recent commit takes into account some of the refactors we discussed today. 

**Items still pending:** removing the ownership checks because Solon will add them into the checkCanAlterToNewOwner method itself; logic tests. 

Currently hitting an error in existing logic tests where a non-existent new role's error is not handled correctly. Instead of "role/user $FAKE_NEW_ROLE does not exist" we get a standard "error while processing" error. I had thought that since we are calling `checkCanAlterToNewOwner` in every ALTER OWNER command, it would output the correct error message. 

Is this because I am now calling the check within the `startExec` method versus the `ReassignOwnedBy` method on the planner — do errors thrown in execution default to "error while processing"?

knz reviewed (commented):
I took the liberty to look at your PR since you were encountering an odd error-related issue. I think I found the cause (see below)

I also reviewed the PR and have a couple of minor nits. I'll  let you process those with the help of your primary reviewer.

> Reviewed 1 of 3 files at r1, 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MHu2-Rr-qvMgX2xhsUd:b-l05ddz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @arulajmani, and @solongordon)

___
*[pkg/sql/alter_database.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu11x8DDBQXWmnBmGu:-MHu11x8DDBQXWmnBmGv:buy6n9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_database.go#L41)):*
> ```Go
> 
> func (n *alterDatabaseOwnerNode) startExec(params runParams) error {
> 	// TODO(angelaw): Remove once Solon's added hasownership to checkcanaltertonewowner
> ```

nit: it's better to preserve capitalization of identifiers even in comments, to ensure that folk who use `grep` to search for references will also find your comment here.

nit: period at the end of sentence
___
*[pkg/sql/alter_database.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu19SL0QKK1kvx2g5s:-MHu19SL0QKK1kvx2g5t:bentm5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_database.go#L64)):*
> ```Go
> 	desc catalog.MutableDescriptor,
> 	newOwner string,
> 	hasOwnership bool,) error{
> ```

I recommend you set up your editor to automatically run `crlfmt` every time a file is saved.
___
*[pkg/sql/alter_schema.go, line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1GRn6UmShaqZzCoZ:-MHu1GRn6UmShaqZzCo_:b-z3p5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_schema.go#L90)):*
> ```Go
> 	jobDescription string,
> ) error {
> 	// TODO(angelaw): Remove once Solon's added hasownership to checkcanaltertonewowner
> ```

ditto above
___
*[pkg/sql/alter_table.go, line 1214 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1K0pAZmVt4g5jRI8:-MHu1K0pAZmVt4g5jRI9:bgly6xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_table.go#L1214)):*
> ```Go
> 
> // Helper method for privilege checking and setting new owner
> // Called in ALTER TABLE and REASSIGN OWNED BY
> ```

The first word of the comment should be the name of the function/method.

nit: periods at the end of sentences.
___
*[pkg/sql/alter_type.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1PpY149P2AzVSxWK:-MHu1PpY149P2AzVSxWL:b-z3p5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_type.go#L300)):*
> ```Go
> func (p *planner) alterTypeOwner(ctx context.Context, n *alterTypeNode, newOwner string) error {
> 	typeDesc := n.desc
> 	// TODO(angelaw): Remove once Solon's added hasownership to checkcanaltertonewowner
> ```

ditto above
___
*[pkg/sql/alter_type.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1R_lCFwcjxWqzWwC:-MHu1R_lCFwcjxWqzWwD:b-l2u14s) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_type.go#L316)):*
> ```Go
> 	}
> 
> 	arrayDesc, err := p.Descriptors().GetMutableTypeVersionByID(ctx, p.txn, typeDesc.ArrayTypeID)
> ```

You forgot to check the `err` here.
___
*[pkg/sql/alter_type.go, line 327 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1TyHBzRd1nHQcp5V:-MHu1TyHBzRd1nHQcp5W:buwdpjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_type.go#L327)):*
> ```Go
> 	ctx context.Context, typeDesc *typedesc.Mutable, newOwner string, hasOwnership bool,
> ) error {
> ```

nit: avoid empty lines as the first line in the body of a function
___
*[pkg/sql/reassign_owned_by.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1a4U0RtTQ7qUzPNR:-MHu1a4U0RtTQ7qUzPNS:b6lw4pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/reassign_owned_by.go#L31)):*
> ```Go
> 
> func (p *planner) ReassignOwnedBy(ctx context.Context, n *tree.ReassignOwnedBy) (planNode, error) {
> 	// Check all roles in old roles exist
> ```

nit: period oat the end of the sentence
___
*[pkg/sql/reassign_owned_by.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1btQ95vCBwIXzRha:-MHu1btQ95vCBwIXzRhb:btfdt37) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/reassign_owned_by.go#L47)):*
> ```Go
> 
> func (n *reassignOwnedByNode) startExec(params runParams) error {
> 	// Search within all descriptors - not limited to this database
> ```

nit: period at the end of the sentence. 

It would also be useful if the comment contained more words. It's not clear what "not limited to this database" means."
___
*[pkg/sql/reassign_owned_by.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1h_MD1toMaeItxXp:-MHu1h_MD1toMaeItxXq:b-i3pbxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/reassign_owned_by.go#L55)):*
> ```Go
> 	lCtx := newInternalLookupCtx(params.ctx, allDescs, nil)
> 
> 	// For each role in old roles
> ```

nit: (here and below): period at the end of sentences.
___
*[pkg/sql/reassign_owned_by.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1qx407kH6STJunYY:-MHu1qx407kH6STJunYZ:b-6p7t4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/reassign_owned_by.go#L81)):*
> ```Go
> 			}
> 		}
> 		for _, schemaDesc := range lCtx.schemaDescs {
> ```

This function is very long. I would encourage you to consider extracting each of these loops into a dedicated sub-function for readability.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
Thanks so much, @knz — very much appreciated your review. I've done the refactoring and fixed that error check. I still seem to be running into the same "error while processing" error, though, in my existing logic tests.

Still leaving a couple of comments/questions up for @solongordon or @arulajmani to take a look at as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MHw3WnD9sgsTN5IZzDB:boz4lq7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @arulajmani, @knz, and @solongordon)

___
*[pkg/sql/alter_database.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu11x8DDBQXWmnBmGu:-MHvyf4o7VtdoKDTdddy:b-10zgjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_database.go#L41)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: it's better to preserve capitalization of identifiers even in comments, to ensure that folk who use `grep` to search for references will also find your comment here.

nit: period at the end of sentence
</blockquote></details>

Thank you! 
___
*[pkg/sql/alter_database.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu19SL0QKK1kvx2g5s:-MHvymy57mW3A1jq7YFA:b-3uhgx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_database.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I recommend you set up your editor to automatically run `crlfmt` every time a file is saved.
</blockquote></details>

Got it :thumbsup: 
___
*[pkg/sql/alter_type.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1R_lCFwcjxWqzWwC:-MHvz_QnBMgJdyWUPi8h:bn6pncs) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_type.go#L316)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

You forgot to check the `err` here.
</blockquote></details>

Great catch, thank you.
___
*[pkg/sql/reassign_owned_by.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlU1BV4E-_97WgsDRd:-MHw3REX4mLFBMEEqFOZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L102)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Will do!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
Updated with further refactors, fixed some parts of logic test and ensuring that errors are called appropriately.

Still working on logic tests (realized late that I could not login as anything other than testuser or testuser2 in the logic tests), but pushing as I wanted to surface a null pointer panic that I get in my first REASSIGN OWNED TO...  logic test that actually has objects to be reassigned. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MHxMWPY6LAUuWHNEmLM:b-s5xpfa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @arulajmani, and @knz)

___
*[pkg/sql/alter_database.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu19SL0QKK1kvx2g5s:-MHxLKVQBxlrbl15GhAo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_database.go#L64)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Got it :thumbsup: 
</blockquote></details>

Done.
___
*[pkg/sql/alter_schema.go, line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1GRn6UmShaqZzCoZ:-MHxLLmlCpBjj4g3YFDX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_schema.go#L90)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto above
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 1214 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1K0pAZmVt4g5jRI8:-MHxLOpW4XT17kQrcv7O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_table.go#L1214)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The first word of the comment should be the name of the function/method.

nit: periods at the end of sentences.
</blockquote></details>

Done.
___
*[pkg/sql/alter_type.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1PpY149P2AzVSxWK:-MHxLPTD-hMaK87QNZzs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_type.go#L300)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto above
</blockquote></details>

Done.
___
*[pkg/sql/alter_type.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1R_lCFwcjxWqzWwC:-MHxLQ1fA1yoVHbeOkWS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/alter_type.go#L316)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Great catch, thank you.
</blockquote></details>

Done.
___
*[pkg/sql/reassign_owned_by.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1qx407kH6STJunYY:-MHxLXw3Du34c8l2UQrP:b-vtio7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/reassign_owned_by.go#L81)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This function is very long. I would encourage you to consider extracting each of these loops into a dedicated sub-function for readability.
</blockquote></details>

I have done so. I'm not sure if it's best practice to make these sub-functions also methods of the reassignOwnedbyNode, but as they need the node itself to write the changes I thought it would be best to do it this way.

Alternatively I could pass in the node via parameters to a method that doesn't have this struct as its receiver. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
For me, this PR is looking good bar solon's additions. should this happen before or after this PR is merged?

> Reviewed 1 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MIM_M0B-SYJDt7aXaeh:bjzbgbj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @arulajmani, @knz, and @otan)

___
*[pkg/sql/alter_database.go, line 65 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIMZxsW4gaKmwreHaR-:-MIMZxsW4gaKmwreHaR0:b-gnxwys) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/alter_database.go#L65)):*
> ```Go
> }
> 
> /* TODO(angelaw): can be potentially re-merged with last method after changes to authorization.go */
> ```

nit: all comments should be of form // TODO(angelaw): instead if block quotes Also usually TODOs come after the comment describing the function.
I usually prefer TODOs to point to tasks -- but I'm guessing this will go away soon
___
*[pkg/sql/reassign_owned_by.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIM_biq8pK3MDzLhXSh:-MIM_biq8pK3MDzLhXSi:bea91i0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/reassign_owned_by.go#L81)):*
> ```Go
> 	}
> 
> 	lCtx := newInternalLookupCtx(params.ctx, allDescs, nil)
> ```

I think this should follow Postgres and only apply changes on the current DB - lest we run into ORM incompatabilities. I know @solongordon advocated for this, so curious for reasoning here.
___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by, line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIMaLGS4M01bDZ3IBjk:-MIMaLGS4M01bDZ3IBjl:b-8lcsvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/logictest/testdata/logic_test/reassign_owned_by#L54)):*
> ```
> 
> statement ok
> CREATE ROLE testuser2 WITH LOGIN
> ```

is owner available in any of the metadata tables?  
seems like there isn't D: (https://github.com/cockroachdb/cockroach/issues/53566), can you add tests to query `pg_namespace` (SHOW SCHEMAS will get metadata from this table, as will the other SHOW COMMANDS get from these respective tables) / pg_type / pg_database / pg_table etc. so that when ownership is added, the author will add these tests automatically later?

something like

````
query TT
SELECT typname, typowner FROM pg_type ORDER BY typname ASC 
```
(note the sort since types can appear out-of-order)
___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by_again, line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIM_YmL3lvmb7-5FpLE:-MIM_YmL3lvmb7-5FpLF:blozn0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/logictest/testdata/logic_test/reassign_owned_by_again#L2)):*
> ```
> statement ok
> CREATE TYPE new_typ AS ENUM ()
> ```

did you intend to include this?


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
I think this should happen before the PR is merged so I can refactor many of the methods to account for the new authorization, and test to see whether anything has broken if using the built-in ALTER authorizations. If time-sensitive I can merge this one and then do the refactoring later. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MIN4wE0-xP2bkmVvNL4:b-ofe9mr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/alter_database.go, line 65 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIMZxsW4gaKmwreHaR-:-MIN2h6v3pP2tIZ3cTPR:b-eupj1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/alter_database.go#L65)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: all comments should be of form // TODO(angelaw): instead if block quotes Also usually TODOs come after the comment describing the function.
I usually prefer TODOs to point to tasks -- but I'm guessing this will go away soon


</blockquote></details>


Fixing this! Thanks for letting me know. I had assumed these would go away before merging the PR but it will depend on the larger question of whether this should be merged before Solon's authorization changes.
___
*[pkg/sql/reassign_owned_by.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlX0KH31xm6ubJT_V_:-MIN5HKE0rTw0vwFEO6v:b-a2ujfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/reassign_owned_by.go#L29)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

A bulk of what you're doing here should happen in the execution phase. 
</blockquote></details>

Changed!
___
*[pkg/sql/reassign_owned_by.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHu1qx407kH6STJunYY:-MIN3vU90f-N1O9MC6IS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/044f2ddefe3f6d8aad8d0f8c93bb250d3b000641/pkg/sql/reassign_owned_by.go#L81)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

I have done so. I'm not sure if it's best practice to make these sub-functions also methods of the reassignOwnedbyNode, but as they need the node itself to write the changes I thought it would be best to do it this way.

Alternatively I could pass in the node via parameters to a method that doesn't have this struct as its receiver. 
</blockquote></details>

Done.
___
*[pkg/sql/reassign_owned_by.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIM_biq8pK3MDzLhXSh:-MIN3hOg1pqI5EpEYbO6:b-yo1u9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/reassign_owned_by.go#L81)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I think this should follow Postgres and only apply changes on the current DB - lest we run into ORM incompatabilities. I know @solongordon advocated for this, so curious for reasoning here.
</blockquote></details>

Hm... I had thought @solongordon said this was because we generally allow for other changes across databases (not just the current one) as well. I don't know much more so happy to entertain both possibilities!
___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by, line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIMaLGS4M01bDZ3IBjk:-MIN3xX4DO7VDUhwrqrk:b-txawtl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/logictest/testdata/logic_test/reassign_owned_by#L54)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is owner available in any of the metadata tables?  
seems like there isn't D: (https://github.com/cockroachdb/cockroach/issues/53566), can you add tests to query `pg_namespace` (SHOW SCHEMAS will get metadata from this table, as will the other SHOW COMMANDS get from these respective tables) / pg_type / pg_database / pg_table etc. so that when ownership is added, the author will add these tests automatically later?

something like

````
query TT
SELECT typname, typowner FROM pg_type ORDER BY typname ASC 
```
(note the sort since types can appear out-of-order)
</blockquote></details>

Not yet, but I know @ajstorm is working on it in #54708. I will add these tests tomorrow. What exactly do the T, I, and B stand for after `query`? 
___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by_again, line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIM_YmL3lvmb7-5FpLE:-MIN5Dsp0WRwYmKtdN1G:b-khfk1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/logictest/testdata/logic_test/reassign_owned_by_again#L2)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

did you intend to include this?
</blockquote></details>

Oops! I did not. Will remove.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by, line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIMaLGS4M01bDZ3IBjk:-MIN7FIh-PRNfhwX50eS:b-dyyom3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/logictest/testdata/logic_test/reassign_owned_by#L54)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Not yet, but I know @ajstorm is working on it in #54708. I will add these tests tomorrow. What exactly do the T, I, and B stand for after `query`? 
</blockquote></details>

Added. The database tests will fail for now until #54982 is merged and rebased.
___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by_again, line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIM_YmL3lvmb7-5FpLE:-MIPlJzk8RqmT6nLWo-s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5cc9ac9a4c2635f5fe2b2e4b6209938d5f4b4f/pkg/sql/logictest/testdata/logic_test/reassign_owned_by_again#L2)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Oops! I did not. Will remove.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

angelapwen commented:
Tests are finished and passing. PR in my opinion is ready pending:

[ ] adding telemetry
[ ] discussion of whether we need to reassign objects for only the current database as Postgres does, or across all databases
[ ] changes to the `authorization.go` file that impact these files and will require refactoring

otan reviewed (commented):
i'll give this another review when solon has his changes in.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MIVuwop4otA_aLC4Wjw:b-tl26dg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/logictest/testdata/logic_test/reassign_owned_by, line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MIVu9yP2-zAYrM_2cRl:-MIVu9yP2-zAYrM_2cRm:bmqquml) ([raw file](https://github.com/cockroachdb/cockroach/blob/5380229d741b8d972b89c6eeb783693aaab7852a/pkg/sql/logictest/testdata/logic_test/reassign_owned_by#L53)):*
> ```
> GRANT testuser TO root
> 
> statement ok
> ```

these test are getting harder to follow because they take multiple lines of setup. can you do something combining multiple statement oks together that set up something together, like so:

```
statement ok
CREATE ROLE testuser2 WITH LOGIN;
GRANT testuser2 TO ROOT

statement ok
CREATE DATABSE d;
ALTER DATABASE d OWNER TO testuser

query TT
SELECT * from pg_data 

....
```

That way multiple lines of setup are easier to read!


<!-- Sent from Reviewable.io -->

angelapwen commented:
I've cleaned up the logic tests and added the tests that check for whether we're reassigning objects only for the current database. Unfortunately it seems that all objects in all databases are still being reassigned (see last section of logic tests). Is there something wrong in the way I'm looking up descriptors in the current database using `newInternalLookupCtx`? It seems as if I should be passing in the current database's descriptor in the `prefix` param to filter.

angelapwen commented:
I have rebased in the changes from #55147 and refactored appropriately. RFAL 🎉

otan reviewed (commented):
All this LGTM, but I think I'm going to let @solongordon give the final stamp as he's been more into the weeds of this than I have been!

> Reviewed 4 of 10 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MJS_2gF94I2EmhTOK23:b-8vmvrt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/authorization.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlNOzX7TApe42ke_fT:-MJSZ93lEJAttubFlJLF:b-l2trhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/authorization.go#L159)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

I chatted with Solon and it sounds like I will be able to rely on the calls to checkCanAlterToNewOwner() nested within each of the Alter * Owner commands, so I may not need this, but just curious:

I had thought I needed to add the planner receiver in order to call IsOwner() directly from sql/reassign_owned_by. Is there a way to access this function elsewhere without adding the receiver? 
</blockquote></details>

I'm assuming Arul meant removing the `p *planner` receiver -- I don't think that needs to be done in this PR. Are you able to do `IsOwner(desc, role)` in reassigned_owned_by.go?
___
*[pkg/sql/resolver.go, line 644 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJSZhfUCIMbOZxJzhrN:-MJSZhfUCIMbOZxJzhrO:b-dwpoc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a281ed9e989e4e0d3bbb34191aac6b4c771909d/pkg/sql/resolver.go#L644)):*
> ```Go
> 	typDescs    map[descpb.ID]*typedesc.Immutable
> 	typIDs      []descpb.ID
> 	schemaIDs   []descpb.ID
> ```

nit: can you move this to line after schemaDescs?


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/authorization.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlNOzX7TApe42ke_fT:-MJS_i1oA_UiN1o8R7wc:b-glhrr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/authorization.go#L159)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I'm assuming Arul meant removing the `p *planner` receiver -- I don't think that needs to be done in this PR. Are you able to do `IsOwner(desc, role)` in reassigned_owned_by.go?
</blockquote></details>

I added the `p *planner` receiver because I wanted to call `IsOwner(desc, role` in reassign_owned_by.go. Without that receiver, I'm unable to access the function in that file. 
___
*[pkg/sql/resolver.go, line 644 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJSZhfUCIMbOZxJzhrN:-MJSaIEg5YyFXFcfCJZW:b4d7ymm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a281ed9e989e4e0d3bbb34191aac6b4c771909d/pkg/sql/resolver.go#L644)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: can you move this to line after schemaDescs?
</blockquote></details>

Will do! Makes sense.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/authorization.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MHlNOzX7TApe42ke_fT:-MJY9Hf-6o4P_gNRVw1s:b761qun) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f4e4dc28af89b21d46452fce163e28f152ec66/pkg/sql/authorization.go#L159)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

I added the `p *planner` receiver because I wanted to call `IsOwner(desc, role` in reassign_owned_by.go. Without that receiver, I'm unable to access the function in that file. 
</blockquote></details>

@arulajmani taught me about public methods in Go 😆 I made the change he requested! (It should be done in this PR because I was the one who added it in)


<!-- Sent from Reviewable.io -->

solongordon reviewed (changes requested):
A few more small comments but then should be good to go!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MJg91Az5Qa0wkfyP2VO:bbbtxcn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/alter_database.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJcxxSR80DL_N7vSyTx:-MJcxxSR80DL_N7vSyTy:b-imh5j5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/alter_database.go#L58)):*
> ```Go
> 	privs := desc.GetPrivileges()
> 
> 	// If the owner we want to set to is the current owner, do a no-op.
> ```

Hmm, it seems like this isn't being treated as a no-op anymore, since `startExec` will still call `writeNonDropDatabaseChange`. I think you could lift this check into `startExec` since it doesn't seem necessary for REASSIGN OWNED BY.
___
*[pkg/sql/alter_schema.go, line 106 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJcyo240k69oDOZ9C2i:-MJcyo240k69oDOZ9C2j:b8pzs21) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/alter_schema.go#L106)):*
> ```Go
> 	privs := scDesc.GetPrivileges()
> 
> 	// If the owner we want to set to is the current owner, do a no-op.
> ```

Same problem as above.
___
*[pkg/sql/alter_table.go, line 1203 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJd-UWo9ZaianAzg3J2:-MJd-UWo9ZaianAzg3J3:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/alter_table.go#L1203)):*
> ```Go
> 	privs := desc.GetPrivileges()
> 
> 	// If the owner we want to set to is the current owner, do a no-op.
> ```

Ditto.
___
*[pkg/sql/reassign_owned_by.go, line 42 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJg1mr3EFR6u0DcO8hA:-MJg1mr3EFR6u0DcO8hB:b-a31wqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/reassign_owned_by.go#L42)):*
> ```Go
> 		}
> 		if !roleExists {
> 			return nil, pgerror.Newf(pgcode.UndefinedObject, "role/user %q does not exist", oldRole)
> ```

Minor: You can just say "role" rather than "role/user".  That's what Postgres does.
___
*[pkg/sql/reassign_owned_by.go, line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJczVsY3Dlp8F1t9fMb:-MJczVsY3Dlp8F1t9fMc:bxtmij7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/reassign_owned_by.go#L102)):*
> ```Go
> 
> func (n *reassignOwnedByNode) reassignDatabaseOwner(
> 	dbDesc *dbdesc.Immutable, oldRole string, params runParams,
> ```

Looks like `oldRole` is unused in this and the following methods.
___
*[pkg/sql/resolver.go, line 770 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJg3PLM8ca7sFSlquqq:-MJg3PLM8ca7sFSlquqr:bmf9zgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/resolver.go#L770)):*
> ```Go
> 		typIDs:      typIDs,
> 		fallback:    fallback,
> 		schemaIDs:   schemaIDs,
> ```

Nit: Move this after `schemaDescs` as well.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)

___
*[pkg/sql/alter_database.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJcxxSR80DL_N7vSyTx:-MJgYBLe-wgsoe1U8Fzm:b5opj9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/alter_database.go#L58)):*
> Hmm, it seems like this isn't being treated as a no-op anymore, since startExec will still call writeNonDropDatabaseChange. I think you could lift this check into startExec since it doesn't seem necessary for REASSIGN OWNED BY.        solongordon (Solon) is:   No elements found. Consider changing the search query. Blocking:
<details><summary><i alt="summary">Quoted 78 lines of code…</i></summary>

>         opposed to resolving the discussion while waiting on somebody else.
>         Dismiss from
>       this discussion
>       •
>       unresolved discussions.
>             
>             This is advanced stuff, you don't need it to get started.You are:  Following: following new comments but neutral on resolving.
>            Discussing:
>             contributing to the discussion but neutral on resolving.
>              Satisfied: in favor of resolving and ending the discussion.
>              Blocking:
>             opposed to resolving the discussion while waiting on somebody else.
>            Working:
>             holding the discussion unresolved while working on something related.
>             No elements found. Consider changing the search query. List is empty. 
>                     done
>                   
>   
>   
>   
>     
>       // If the owner we want to set to is the current owner, do a no-op.
>     
>     
>     
>   
> 
> 
>   
>     
>       if newOwner == privs.Owner {
>     
>     
>     
>   
>   
>   
>     
>       if newOwner == privs.Owner {
>     
>     
>     
>   
> 
> 
>   
>     
>       return nil
>     
>     
>     
>   
>   
>   
>     
>       return nil
>     
>     
>     
>   
> 
> 
>   
>     
>       }
>     
>     
>     
>   
>   
>   
>     
>       }
>     
>     
>     
>   
> 
</details>
> EXPANDEXPAND
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

>                 Collapsed:
>               18 lines
>             with
>             
>               no changes
>             
>       single
>     
>       auto
>     130
</details>
Ah, yeah that's right. Thanks for catching that. I'll lift all of these no-ops into startExec. You didn't mark it on the alter_type.go file but I've lifted it there as well as it originally had the same structure. 
___
*[pkg/sql/alter_schema.go, line 106 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJcyo240k69oDOZ9C2i:-MJgYcgF5U1DicYpi-p8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/alter_schema.go#L106)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Same problem as above.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table.go, line 1203 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJd-UWo9ZaianAzg3J2:-MJgZqns7aFdjwd2u-N5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/alter_table.go#L1203)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.
___
*[pkg/sql/reassign_owned_by.go, line 42 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJg1mr3EFR6u0DcO8hA:-MJg_53v6mejiQo_uKqQ:bqzjni6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/reassign_owned_by.go#L42)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Minor: You can just say "role" rather than "role/user".  That's what Postgres does.
</blockquote></details>

Discussed on Slack — keeping this to "role/user" as this is the language already used in existing checks. 
___
*[pkg/sql/reassign_owned_by.go, line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-MJczVsY3Dlp8F1t9fMb:-MJg_Zff8U08XrHmzJF5:b-5karrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b325ab4112472683240a377cfdd3250c6af25b/pkg/sql/reassign_owned_by.go#L102)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Looks like `oldRole` is unused in this and the following methods.
</blockquote></details>

Thank you, got it 


<!-- Sent from Reviewable.io -->

solongordon reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 7 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54594#-:-MJgnom3DSJ6sK6ZfmIX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @angelapwen, @arulajmani, @knz, @otan, and @solongordon)



<!-- Sent from Reviewable.io -->

angelapwen commented:
bors r=solongordon
